File: HW09/a00824566.hs
-- ====================
-- Complete the following functions and submit your file to Canvas.
-- ====================
-- Do not change the names of the functions. 
-- Do not change the number of arguments in the functions.
-- If your file cannot be loaded by the Haskell compiler, your submission may be cancelled. 
-- Then, submit only code that works.
-- ====================
-- Grading instructions:
-- There is a series of test cases for each function. In order to state that your function
-- "works as described", your output must be similar to the expected one in each case.

-- === invert ===

invert :: [t] -> [t]
invert [] = []
invert lst = last lst : invert (init lst)

-- === or ===

listor :: [Int] -> [Int] -> [Int]
listor [] [] = []
listor (a : b) (x : y) = if a == x
 then if a == 0
 then 0 : listor b y
 else 1 : listor b y
 else 1 : listor b y

-- === multiples ===

multiples :: [Int] -> Int -> [Int]
multiples [] _ = []
multiples (a : b) x = if mod a x == 0
 then a : multiples b x
 else multiples b x

-- === differences ===

diff :: Int -> [Int] -> [Int]
diff _ [] = []
diff x (a : b) = if null b
 then abs(x - a) : diff x b
 else abs(a - head b) : diff x b

differences :: [Int] -> [Int]
differences lst = diff (head lst) lst

-- === toBinaryString ===

toBin :: Int -> [Char]
toBin 0 = ['0']
toBin 1 = ['1']
toBin x = if mod x 2 == 1
 then '1' : toBin(div x 2)
 else '0' : toBin(div x 2)

toBinaryString :: Int -> [Char]
toBinaryString lst = reverse(toBin lst)

-- === modulo ===

modulo :: Int -> Int -> Int
modulo 0 _ = 0 
modulo x y = let w = x - y in if w >= y
 then modulo w y
 else w

-- === evaluate ===

evaluate :: [Double] -> Double -> Double
evaluate [] _ = 0
evaluate (a : b) x = let y = length b in a * x ^^ y + evaluate b x 

-- === cleanString ===

unique :: Char -> [Char] -> [Char]
unique _ [] = []
unique c (a : b) = if c /= a
 then a : unique a b
 else unique a b

cleanString :: [Char] -> [Char]
cleanString [] = []
cleanString (a : b) = a : unique a b

-- === iSort ===

insert :: Int -> [Int] -> [Int]
insert x [] = [x]
insert x (a : b) = if x < a
 then x : a : b
 else a : (insert x b)

iSort :: [Int] -> [Int]
iSort [] = []
iSort (a : b) = insert a (iSort b)

-- === Test cases ===

main = do 
    print "=== invert ==="
    print $ invert ([] :: [Int])-- []
    print $ invert [1, 2, 3, 4, 5] -- [5,4,3,2,1]
    print $ invert "hello world!" -- "!dlrow olleh"
    print "=== listor ==="
    print $ listor [1, 1, 0] [0, 1, 0] -- [1,1,0]
    print $ listor [1, 0, 1, 0] [0, 0, 1, 1] -- [1,0,1,1]
    print $ listor [1, 0, 1, 0, 1] [1, 1, 1, 0, 0] -- [1,1,1,0,1]
    print "=== multiples ==="
    print $ multiples [2, 4, 5, 6] 2 -- [2,4,6]
    print $ multiples [9, 27, 8, 15, 4] 3 -- [9,27,15]
    print $ multiples [9, 8, 17, 5] 6 -- []
    print "=== differences ==="
    print $ differences [1, 2, 4, 8, 20] -- [1,2,4,12,19]
    print $ differences [5, 9, 13, 27, 100, 91, 4] -- [4,4,14,73,9,87,1]
    print $ differences [99] -- [0]
    print $ differences [] -- [] 
    print "=== toBinaryString ==="
    print $ toBinaryString 0 -- "0"
    print $ toBinaryString 1 -- "1"
    print $ toBinaryString 7 -- "111"
    print $ toBinaryString 32 -- "100000"
    print $ toBinaryString 1024 -- "10000000000"
    print "=== modulo ==="
    print $ modulo 10 2 -- 0
    print $ modulo 15 4 -- 3
    print $ modulo 20 9 -- 2
    print $ modulo 77 10 -- 7
    print "=== evaluate ==="
    print $ evaluate ([] :: [Double]) 100 -- 0.0
    print $ evaluate [2, 3.1, 10, 0] 2 -- 48.4
    print $ evaluate [10, 0] 2 -- 20.0
    print $ evaluate [1, 2, 3, 4, 5] 3 -- 179.0
    print "=== cleanString ==="
    print $ cleanString ([] :: String) -- ""
    print $ cleanString "yyzzza" -- "yza"
    print $ cleanString "aaaabbbccd" -- "abcd"
    print "=== iSort ==="
    print $ iSort [] -- []
    print $ iSort [1] -- [1]
    print $ iSort [1, 6, 3, 10, 2, 14] -- [1,2,3,6,10,14]

File: CH08/a00824566.erl
-module(challenge08).
-export([test01/0, increase/0, test02/0, dynamicIncrease/1, test03/0, calculator/1, test04/0, tictac/2]).

% ====================
% Complete the following functions and submit your file to Canvas.
% ====================
% Do not change the names of the functions. 
% Do not change the number of arguments in the functions.
% If your file cannot be loaded by the Erlang compiler, your submission may be cancelled. 
% Then, submit only code that works.
% ====================
% Grading instructions:
% There is a series of test cases for each function. In order to state that your function
% "works as described", your output must be similar to the expected one in each case.

% === increase ===

increase() ->
	receive
		X when is_number(X) -> io:format(">>~p~n", [X + 1]), increase();
		_ -> ok
	end.

test01() -> 
	io:format("=== increase ===~n"),
	Pid = spawn(fun challenge08:increase/0),
	Pid ! 5, 	% >> 6 is printed on screen.
	Pid ! 10, 	% >> 11 is printed on screen.
	Pid ! 99, 	% >> 100 is printed on screen.
	Pid ! bye, 	% The process ends.
	Pid ! 5, 	% Nothing happens since the process is already done.
	ok.			% This is the return value for test01 (it will eventually be printed on screen).

% === dynamicIncrease ===

dynamicIncrease(Y) -> receive
		X when is_number(X) -> io:format(">>~p~n", [X + Y]), dynamicIncrease(Y + 1);
		_ -> ok
	end.

test02() -> 
	io:format("=== dynamicIncrease ===~n"),
	Pid = spawn(challenge08, dynamicIncrease, [1]),
	Pid ! 5, 	% >> 6 is printed on screen.
	Pid ! 10, 	% >> 12 is printed on screen.
	Pid ! 99, 	% >> 102 is printed on screen.
	Pid ! 1000, % >> 1004 is printed on screen.
	Pid ! bye, 	% The process ends.
	Pid ! 5, 	% Nothing happens since the process is already done.
	ok.			% This is the return value for test02 (it will eventually be printed on screen).

% === calculator ===

calculator(Y) -> 
	receive
		{xadd, X} -> calculator(Y + X);
		{xmul, X} -> calculator(Y * X);
		{xsub, X} -> calculator(Y - X);
		{xdiv, X} -> calculator(Y / X);
		finish -> io:format(">>~p~n", [Y]), ok	
	end.

test03() -> 
	io:format("=== calculator ===~n"),
	Pid = spawn(challenge08, calculator, [0]),
	Pid ! {xadd, 10},	% Nothing is printed on screen.
	Pid ! {xmul, 5}, 	% Nothing is printed on screen.
	Pid ! {xsub, 3}, 	% Nothing is printed on screen.
	Pid ! finish, 		% 47 is printed on screen.
	Pid ! bye, 			% Nothing happens since the process is already done.
	Pid ! 5, 			% Nothing happens since the process is already done.
	ok.					% This is the return value for test03 (it will eventually be printed on screen).

% === tictac ===

tictac(Value, Num) ->
	receive
		{start, Pid} -> io:format("~p~n", [Value])
	end,
	receive
		after 1000 ->
			Pid ! {start, self()}, if
				Num > 0 -> tictac(Value, Num - 1);
				true ->	ok
		end
	end.

test04() ->
	io:format("=== tictac ===~n"),
	PidTic = spawn(challenge08, tictac, [tic, 10]),
	io:format("~p~n", [PidTic]),
	PidTac = spawn(challenge08, tictac, [tac, 10]),
	PidTic ! {start, PidTac},
	ok.
